File: Data Structures & Algorithms/linked-list-cycle-detection/submission-0.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */

class Solution {
    public boolean hasCycle(ListNode head) {
        // 1 -> 2
        // 0 -> 1
        Map<ListNode, Integer> map = new HashMap<>();

        while (head != null) {
            map.put(head, map.getOrDefault(head, 0) + 1);
            if (map.get(head) > 1) {
                return true;
            }

            head = head.next;
        }

        return false;
    }
}

File: Data Structures & Algorithms/remove-node-from-end-of-linked-list/submission-0.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next

class Solution:
    def removeNthFromEnd(self, head: Optional[ListNode], n: int) -> Optional[ListNode]:
        # 1 2 3 4 
        length = 0
        current = head
        while current:
            length += 1
            current = current.next
        
        prevIndex = length - n - 1
        if prevIndex < 0:
            if head.next:
                head = head.next
                return head
            return None

        prev = head
        for i in range(prevIndex):
            prev = prev.next

        temp = prev.next.next
        prev.next = temp 

        return head 



        